[MAJOR] v2 rewrite, removes SimpleSAMLphp for onelogin/php-saml

madmatt opened the pull request:
This is a major update to the module that now uses HTTP-Post binding instead of HTTP-Artifact binding to communicate with RealMe, which makes the codebase a lot cleaner, requires less hacks, and makes it a lot easier to use.

Some further work is likely to be required to complete this entirely, but this is now being used stable by a couple of sites, so it's time to merge it in and get a little more usage.

dhensby commented on composer.json:
```diff
@@ -18,15 +18,12 @@
         "composer/installers": "*",
         "silverstripe/framework": "~3.1",
-        "madmatt/simplesamlphp": "dev-ss-master"
+        "onelogin/php-saml": "v2.10.2"
```
I'd urge against locking to a single version - this makes taking in security updates from the upstream harder and can be frustrating if you rely on a bug fix in a subsequent version.

A project maintainer should be given the decision to lock onto a version, rather than the module.

madmatt replied:
Updated to lock on `~2.10` instead, though somewhat irrelevant as we currently have to require people set a custom repo for that anyway, due to changes that can't (yet) be merged back into mainline... working on that separately.

dhensby commented on .gitignore:
```diff
@@ -1,5 +1 @@
-.profile
-vendor/*
-.sass-cache/*
+/.idea
```
IDE specific ignores should live in a local global .gitignore on your machine

dhensby commented on README.md:
````diff
@@ -145,9 +151,24 @@ In this environment, RealMe provide all SSL certificates required to communicate
+}
+```
+
+- You can also access user data from any descendant of SiteTree using `$RealMeUser`
````
Is this referring to the templates or PHP code? If it's templates then you should make `$RealMeUser` provided by a `TemplateGlobalProvider` so it's available on all templates. Otherwise this needs clarity as it's not clear how one would access `$RealMeUser` in `SiteTree` descendants...

dhensby commented on code/RealMeAuthenticator.php:
```diff
@@ -25,51 +25,26 @@ protected static function on_register()
 
         $cacheKey = 'RegisterCheck';
-        if (true === (bool) $cache->load($cacheKey)) {
+        if ((bool) $cache->load($cacheKey)) {
```
You can use `$cache->test($cacheKey)` to check a cache exists - probably better than loading the whole cached data just to convert to a bool

jakxnz replied:
@dhensby The cache content is a `boolean` represented as a `string`...

Line 51:
```
$cache->save('true', $cacheKey);
```

Testing for availability is not precisely congruent with the intended logic, I don't think. Do you have any thoughts on that?

dhensby replied:
Ah, didn't spot that. Howver, if you want false you store `'false'` this would cast to true too

madmatt replied:
Nice catch, thanks. This is part of the original module. Have taken Jackson's changes for this one

dhensby commented on code/RealMeService.php:
```diff
@@ -205,58 +233,189 @@ class RealMeService extends Object
+            $errors = $this->getAuth()->getErrors();
+            $translatedMessage = null;
+
+            if(is_array($errors) && sizeof($errors) > 0) {
```
`!empty()` is probably more efficient than counting `$errors` and evaluating against `>0`

dhensby commented on code/RealMeService.php:
```diff
@@ -205,58 +233,189 @@ class RealMeService extends Object
+            die;
+        }
+
+        $this->config()->user_data = $authData;
```
Storing this data in the `Config` system doesn't seem like the right place for it. What are you trying to achieve by doing so?

dhensby commented on code/RealMeService.php:
```diff
@@ -320,205 +465,291 @@ public function getBackURL()
-        return (defined('REALME_CERT_DIR') ? REALME_CERT_DIR : null);
+        $cfg = $this->config();
+        $name = $this->getConfigurationVarByEnv('idp_x509_cert_filenames', $cfg->realme_env, $cfg->integration_type);
+        return Controller::join_links($this->getCertDir(), $name);
```
How well does this work for joining file paths? I'm not sure that's the intended use... I'm thinking windows paths might not work so well

dhensby replied:
@tractorcow says not to use `Controller` for anything filesystem related - use `realpath` and join them with vanilla string concatenation

dhensby commented on code/RealMeService.php:
```diff
@@ -320,205 +465,291 @@ public function getBackURL()
-            if (!$strongSalt || !$strongPassword || !$salt || !$password) {
-                $password = null; // Ensure the password is strong, return null if we can't guarantee a strong one
+            // If the file does not contain any header information and the content type is certificate, just return it
+            if($contentType == 'certificate' && !preg_match('/-----BEGIN/', $certificateContents)) {
```
this method feels icky - would it be better to use a library - perhaps https://github.com/phpseclib/phpseclib?

jakxnz replied:
@dhensby To help us get this pull request through, it might be worthwhile retro-actively raising this as an issue with a dedicated pull request. Would you be ok with that?

dhensby replied:
sure

madmatt replied:
Yeah, I don't like this either, but I couldn't find a decent certificate parsing lib (thought I didn't try phpseclib, didn't notice the x509_load option) that handled malformed certs like this (which RealMe provides for MTS). Will raise an issue for it.

dhensby commented on code/exceptions/RealMeException.php:
```diff
@@ -0,0 +1,12 @@
+<?php
+class RealMeException extends Exception
+{
+    const INVALID_IDENTITY_VALUE = 0;
```
Would it be better to have subclasses for exception types rather than using constants?

jakxnz replied:
@dhensby subclasses feels a bit overkill just for some structure around status codes, do you have a standard we could refer to?

dhensby replied:
Using sublcasses is more expressive and clearer what's going on IMO... I'm not sure there's any documented standard on exception classes vs codes but it's pretty normal in popular libraries for there to be different exception classes - I've not really ever seen this approach

dhensby commented on docs/en/installation.md:
```diff
@@ -1,56 +1,24 @@
 
-The module is best installed via Composer by running the below command:
+The module is best installed via Composer, by adding the below to your composer.json. For now, we need to specify a 
+custom version of the excellent onelogin/php-saml module to fix some XMLDSig validation errors with the RealMe XML 
```
is this still the case?

madmatt replied:
Unfortunately yes. I haven't had a chance to properly fix this yet, and the upstream module author wasn't able to solve this without more time from me.

dhensby commented on tests/RealMeServiceTest.php:
```diff
@@ -0,0 +1,78 @@
+        $this->pathForTempCertificate = ASSETS_PATH . '/tmpcert.pem';
+
+        // Test standard certificate
+        $contents = <<<EOF
```
I'd probably keep these test files as actual files in the test dir then copy them to where they need to go. Leaving them inline like this feels a bit fragile

jakxnz commented on code/extensions/RealMeSecurityExtension.php:
```diff
@@ -91,17 +94,50 @@ public function realme()
+
+                // If more session vars are set here, they must be cleared in realmeLogout()
+                Session::set('RealMeSessionData', serialize($authData));
+                Session::set('RealMeOriginalResponse', $_POST['SAMLResponse']);
```
@dhensby Do you have any suggestions as to what context to sanatise for?

In the current build, it seems like this was assigned to the session for posterity. Might it be worthwhile simply naming the reference more appropriately, so that users can sanatise for the correct case when using the SAMLResponse? e.g `Session::set('RealMe.RawSAMLResponse', $_POST['SAMLResponse']);`

dhensby replied:
That may be suitable, I'm just very weary of storing raw external data in sessions. Most users assume session data is safe

madmatt replied:
Fair point, but this is set only after we determine that the user is authenticated via RealMe. So to be malicious, you would have to fake a `SAMLResponse` that is both malicious *and* still passes all the SAML checks (including XML signature validation). So basically, this should be impossible.

Then again, just in case, we can just run `Convert::raw2xml()` over it, as the value saved here is just a base64-encoded string anyway.

jakxnz commented:
I have a **near complete** set of commits to make to this branch, which cover @dhensby's peer review, for all items except "certificate handling" (https://github.com/silverstripe/silverstripe-realme/pull/19#discussion_r107585704).

The incomplete branch is found here: https://github.com/silverstripe-terraformers/silverstripe-realme/tree/pulls/onelogin

It's awaiting the ability to test the changes over MTS. My access to RealMe is interrupted at present.

madmatt commented:
Thanks for all your changes @jakxnz, they're great. I've reverted the change to installation.md as noted above but otherwise accepted them all, and squashed them into my fork branch and updated this PR as noted. Have also raised #20 as noted in the comments above on this PR.

dhensby commented:
@madmatt Let me know if you want any more reviews :)

caffeineinc commented:
@dhensby I'm taking it up on @madmatt 's behalf, since he's gone on leave for another 57 days! 

Is there anything left todo, it looks like most changes are sorted here?

dhensby commented:
@caffeineinc that makes sense then ;)

It doesn't look like anything significant has changed since my last review, so I'm happy. I can't claim to be an expert on realme/SAML so my feedback has really been more around style/convention and less about solving the actual problem.

I am worried about using custom forks becuase it means every project that uses this module needs to modify their root composer.json file to bring it in... but I suppose we don't have much choice apart from pushing the upstream maintainers a little.

dhensby commented:
@caffeineinc if you've now got a custom fork then you'll need to update the docs
